File: alpha/experiments/item-storage-protocol-standalone/src/exp/e9_grace.ts
// E9. The grace ledger — mercy is represented in the books, not off-book. Grace
// events are first-class signed ledger entries: a fee waiver with a reason code, a
// deceased-member hold (rent accrues to the co-op's own grace account for a fixed
// term), and throttle-instead-of-cutoff during a payment lapse. Each is a forward
// entry that nets to zero against the co-op grace account, so the books still
// balance and grace totals are reportable per period.

import type { World } from "../world.ts";
import { Checker, type ExperimentResult } from "../types.ts";
import { rentCents } from "../pricing.ts";

const PERIOD = 30;

export function run(world: World): ExperimentResult {
  const c = new Checker();
  const { customer, provider, manifest } = world;

  let customerCreditsCents = 0; // negative: credits applied to members' bills
  const perPeriod: (string | number)[][] = [];

  // Emit a signed grace event and its offsetting entry into the co-op grace
  // account. The customer credit (-amount) and the grace-account charge (+amount)
  // net to zero by construction.
  function graceEvent(reasonCode: string, amountCents: number, note: string): void {
    customerCreditsCents -= amountCents;
    world.graceAccountCents += amountCents;
    provider.ledger.append("grace-event", world.clock.iso(), {
      reasonCode, amountCents, note,
      customerCredit: -amountCents, graceAccountCharge: amountCents,
      graceAccountBalance: world.graceAccountCents,
    }, [provider.signer(), customer.signer()]);
  }

  function periodRent(start: number, end: number): number {
    return rentCents(world.byteDays(start, end));
  }

  // --- Scenario 1: a fee waiver. ---
  {
    const start = world.clock.now();
    world.clock.advanceDays(PERIOD);
    const end = world.clock.now();
    const rent = periodRent(start, end);
    const fee = 25; // a one-off service fee...
    graceEvent("FIRST_TIME_HARDSHIP", fee, "waived a late fee, once, because we could afford to");
    const stmt = world.commitStatement({
      periodStartDay: start, periodEndDay: end,
      openingRoot: manifest!.root, closingRoot: manifest!.root, byteDays: world.byteDays(start, end),
      rentCents: rent, postageBytes: 0, postageCents: 0,
      auditCount: 0, auditBytes: 0, auditCents: 0, auditTier: "none",
      graceCents: -fee, feesCents: fee, totalCents: rent + fee - fee,
    });
    perPeriod.push([stmt.period, "fee-waiver", rent, fee, -fee, stmt.totalCents]);
    c.eq("fee waiver: member's total is rent only (fee fully waived)", stmt.totalCents, rent);
  }

  // --- Scenario 2: deceased-member hold, rent to the co-op for a fixed term. ---
  for (let held = 0; held < 3; held++) {
    const start = world.clock.now();
    world.clock.advanceDays(PERIOD);
    const end = world.clock.now();
    const rent = periodRent(start, end);
    graceEvent("DECEASED_MEMBER_HOLD", rent, `estate hold period ${held + 1}/3: rent carried by the co-op`);
    const stmt = world.commitStatement({
      periodStartDay: start, periodEndDay: end,
      openingRoot: manifest!.root, closingRoot: manifest!.root, byteDays: world.byteDays(start, end),
      rentCents: rent, postageBytes: 0, postageCents: 0,
      auditCount: 0, auditBytes: 0, auditCents: 0, auditTier: "none",
      graceCents: -rent, feesCents: 0, totalCents: 0,
    });
    perPeriod.push([stmt.period, "deceased-hold", rent, 0, -rent, stmt.totalCents]);
    c.eq(`deceased hold period ${held + 1}: estate owes nothing`, stmt.totalCents, 0);
  }

  // --- Scenario 3: throttle-instead-of-cutoff during a payment lapse. ---
  {
    const start = world.clock.now();
    world.clock.advanceDays(PERIOD);
    const end = world.clock.now();
    const rent = periodRent(start, end);
    graceEvent("PAYMENT_LAPSE_THROTTLE", rent, "service throttled, not cut off; rent carried this period");
    const stmt = world.commitStatement({
      periodStartDay: start, periodEndDay: end,
      openingRoot: manifest!.root, closingRoot: manifest!.root, byteDays: world.byteDays(start, end),
      rentCents: rent, postageBytes: 0, postageCents: 0,
      auditCount: 0, auditBytes: 0, auditCents: 0, auditTier: "none",
      graceCents: -rent, feesCents: 0, totalCents: 0,
    });
    perPeriod.push([stmt.period, "throttle", rent, 0, -rent, stmt.totalCents]);
    c.ok("throttle: service continues and the shortfall is carried, not lost", stmt.totalCents === 0);
  }

  // Every grace event nets to zero against the co-op grace account.
  c.eq("grace events net to zero against the co-op grace account",
    customerCreditsCents + world.graceAccountCents, 0);

  // Grace totals are reportable per period (we have a per-period table).
  c.ok("grace totals are reportable per period", perPeriod.length === 5);

  // No grace event edits history: all are forward entries; the provider ledger
  // still verifies as an unbroken append-only chain.
  c.ok("all grace events are forward entries (no history edited)",
    provider.ledger.entries.filter((e) => e.kind === "grace-event").length === 5);

  return {
    id: "E9",
    title: "The grace ledger",
    plainSentence: "The receipts make fairness legible; the margin makes mercy affordable.",
    assertions: c.results,
    tables: [
      {
        title: "Grace events (cents)",
        headers: ["period", "kind", "rent", "fee", "grace credit", "member total"],
        rows: perPeriod,
      },
    ],
    notes: [`Co-op grace account absorbed ${world.graceAccountCents} cents this run.`],
  };
}

File: alpha/experiments/item-store/tests/e9_grace.rs
//! Phase 6 wiring test — ports E9 (the grace ledger) from the oracle
//! (`src/exp/e9_grace.ts`): grace is represented in the books, not off-book.
//! Grace events are first-class co-signed ledger entries (fee waiver, deceased-
//! member hold, throttle-instead-of-cutoff); each nets to zero against the co-op
//! grace account; none edits history (all forward entries, so the provider
//! ledger still verifies unbroken); grace totals are reportable per event; and a
//! waived fee lands in the period statement as rent only.

use std::collections::BTreeMap;

use item_store::crypto::derive_keypair;
use item_store::grace::{GraceAccount, GRACE_EVENT_KIND};
use item_store::identity::derive_id;
use item_store::ledger::{verify_entries, Ledger, Signer};
use item_store::statements::{build_statement, verify_chain, StatementBody, GENESIS_STATEMENT};

#[test]
fn grace_events_are_cosigned_forward_entries_that_net_to_zero() {
    let provider = derive_keypair("e9", "provider");
    let provider_id = derive_id(&provider.verifying_key());
    let customer = derive_keypair("e9", "customer");
    let customer_id = derive_id(&customer.verifying_key());

    let mut ledger = Ledger::new(&provider_id);
    let mut grace = GraceAccount::new();

    // Scenario 1: a one-off fee waiver.
    let ev = grace.record_event(
        &mut ledger,
        "t1",
        "FIRST_TIME_HARDSHIP",
        25,
        "waived a late fee, once, because we could afford to",
        &[
            Signer::new(&provider_id, &provider),
            Signer::new(&customer_id, &customer),
        ],
    );
    assert_eq!(ev.customer_credit_cents, -25);
    assert_eq!(ev.grace_account_charge_cents, 25);
    assert_eq!(ev.grace_account_balance_cents, 25);
    assert_eq!(grace.balance_cents() + grace.customer_credits_cents(), 0);

    // Scenario 2: deceased-member hold — rent carried by the co-op, three periods.
    for held in 0..3 {
        let ev = grace.record_event(
            &mut ledger,
            "t2",
            "DECEASED_MEMBER_HOLD",
            40,
            &format!("estate hold period {}/3", held + 1),
            &[
                Signer::new(&provider_id, &provider),
                Signer::new(&customer_id, &customer),
            ],
        );
        assert_eq!(ev.customer_credit_cents, -40);
        assert_eq!(grace.balance_cents() + grace.customer_credits_cents(), 0);
    }

    // Scenario 3: throttle-instead-of-cutoff during a payment lapse.
    grace.record_event(
        &mut ledger,
        "t3",
        "PAYMENT_LAPSE_THROTTLE",
        40,
        "service throttled, not cut off; rent carried this period",
        &[
            Signer::new(&provider_id, &provider),
            Signer::new(&customer_id, &customer),
        ],
    );

    // Five forward grace entries; totals reportable; credits and charges balance.
    let grace_entries = ledger
        .entries()
        .iter()
        .filter(|e| e.kind == GRACE_EVENT_KIND)
        .count();
    assert_eq!(
        grace_entries, 5,
        "grace events are reportable (all forward entries)"
    );
    assert_eq!(
        grace.balance_cents() + grace.customer_credits_cents(),
        0,
        "credits and charges net to zero",
    );
    assert_eq!(grace.balance_cents(), 25 + 40 * 3 + 40);
    assert_eq!(grace.customer_credits_cents(), -(25 + 40 * 3 + 40));

    // Co-signed by both parties, the append-only provider ledger verifies unbroken.
    let ring: BTreeMap<String, _> = [
        (provider_id.clone(), provider.verifying_key()),
        (customer_id.clone(), customer.verifying_key()),
    ]
    .into_iter()
    .collect();
    assert!(
        verify_entries(ledger.entries(), &ring).is_empty(),
        "the co-signed grace ledger verifies as an unbroken append-only chain",
    );
}

#[test]
fn a_waived_fee_lands_in_the_statement_as_rent_only() {
    // fees_cents = fee, grace_cents = -fee (a credit), so the member's total is
    // rent only; the statement still hashes and chains.
    let rent: u64 = 12;
    let fee: u64 = 25;
    let body = StatementBody {
        period: 0,
        period_start_day: 0,
        period_end_day: 30,
        opening_root: "r".to_owned(),
        closing_root: "r".to_owned(),
        byte_days: 0,
        rent_cents: rent,
        postage_bytes: 0,
        postage_cents: 0,
        audit_count: 0,
        audit_bytes: 0,
        audit_cents: 0,
        audit_tier: "none".to_owned(),
        grace_cents: -i64::try_from(fee).expect("fee fits i64"),
        fees_cents: fee,
        total_cents: rent, // rent + fee - fee
        prev_statement_hash: GENESIS_STATEMENT.to_owned(),
    };
    let stmt = build_statement(body);
    assert_eq!(
        stmt.body().grace_cents,
        -25,
        "the waiver is a negative grace credit"
    );
    assert_eq!(stmt.body().total_cents, rent, "member pays rent only");
    assert!(
        verify_chain(std::slice::from_ref(&stmt)).is_ok(),
        "the statement chains"
    );
}
